python/device/device.py

```diff
@@ -108,39 +115,100 @@ def __init__(self):
         self.new_pingback_event = threading.Event()
         # for telemetry
         self.telemetry_queue = queue.Queue()
+        # for metrics
+        self.reporter = MetricsReporter()
+        self._configure_azure_monitor_metrics()
+
+    def _configure_azure_monitor_metrics(self):
+        self.reporter.add_float_measurement(
+            "process_cpu_percent",
+            "processCpuPercent",
+            "Amount of CPU usage by the process",
+            "percentage",
+        )
+        self.reporter.add_integer_measurement(
+            "process_working_set",
+            "processWorkingSet",
+            "All physical memory used by the process",
+            "bytes",
+        )
+        self.reporter.add_integer_measurement(
+            "process_bytes_in_all_heaps",
+            "processBytesInAllHeaps",
+            "All virtual memory used by the process",
+            "bytes",
+        )
+        self.reporter.add_integer_measurement(
+            "process_private_bytes",
+            "processPrivateBytes",
+            "Amount of non-shared physical memory used by the process",
+            "bytes",
+        )
+        self.reporter.add_integer_measurement(
+            "process_working_set_private",
+            "processWorkingSetPrivate",
+            "Amount of non-shared physical memory used by the process",
+            "bytes",
+        )
+        self.reporter.add_integer_measurement(
+            "total_messages_sent",
+            "totalMessagesSent",
+            "count of messages sent and ack'd by the transport",
+            "message(s)",
+        )
+        self.reporter.add_integer_measurement(
+            "total_messages_acknowledged",
+            "totalMessagesAcknowledged",
+            "count of messages sent to iothub with receipt verified via service sdk",
+            "message(s)",
+        )
+        self.reporter.add_integer_measurement(
+            "message_backlog",
+            "messageBacklog",
+            "count of messages waiting to be sent",
+            "message(s)",
+        )
+        self.reporter.add_integer_measurement(
+            "messages_in_flight",
+            "messagesInFlight",
+            "count of messages sent to iothub but not ack'd by the transport",
+            "message(s)",
+        )
+        self.reporter.add_integer_measurement(
+            "messages_unacknowledged",
+            "messagesUnacknowledged",
+            "count of messages sent to iothub and acked by the transport, but receipt not verified via service sdk",
```

Comment:
I'm confused, there are 2 levels of acknowledgment, transport and service?

FWIW, we don't seem to have this intermediate stage in .NET or Java, at least how the API is designed.

Reply:
In my next PR, I've renamed these to acked and ReceeivedByService.

---
In reply to: [505741511](https://github.com/Azure/iot-sdk-longhaul/pull/15#discussion_r505741511) [](ancestors = 505741511)